cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java

```diff
@@ -101,10 +101,14 @@ CloseableClassLoader createClassLoader(File unpackDir) {
     }
 
     final ClassLoader finalProgramClassLoader = programClassLoader;
+    final ClassLoader finalSparkClassLoader = sparkClassLoader;
     return new CloseableClassLoader(programClassLoader, () -> {
       if (finalProgramClassLoader instanceof Closeable) {
         Closeables.closeQuietly((Closeable) finalProgramClassLoader);
       }
+      if (finalSparkClassLoader instanceof Closeable) {
+        Closeables.closeQuietly((Closeable) finalSparkClassLoader);
```

Comment:
We shouldn't be closing the spark classloader here, as the spark classloader is shared among all artifacts that are  used for the same program run.

Reply:
Discussed offline, there are two separate code paths for creating the spark class loader when executing programs and when loading artifacts. When artifacts are loaded, the class loader is not being re-used. https://github.com/cdapio/cdap/pull/15483 will ensure that the spark class loader is re-used in appfabric, but it will not ensure re-use in task workers since the Class Loader factory can't be shared across runnable tasks due to the API of RunnableTasks. In next release we will ensure re-use for task workers and appfabric together. For now, we just close the spark class loaders to free memory.